src/name.rs

```diff
@@ -77,20 +77,28 @@ impl<'a> From<DNSNameRef<'a>> for DNSName {
 #[derive(Clone, Copy)]
 pub struct DNSNameRef<'a>(untrusted::Input<'a>);
 
+/// A marker struct indicating that a `DNSNameRef` could not built from the provided input.
+///
+/// This struct is voluntarily kept separate from the `Error` enum to emulate the behavior of the
+/// still unstable `TryFrom` trait, leaving the calling code free to take corrective action or
+/// convert `NotADNSName` into an error type it knows about.
+#[derive(Clone, Copy)]
+pub struct NotADNSName<'a>(untrusted::Input<'a>);
```

Comment:
Do you think the error needs to include a reference to the invalid input? Usually you would have lifetime errors with this code in practice, I imagine. I think we should just make it `pub struct NotADNSName(())`.

Reply:
Should this type have a name that ends in an `Error` suffix?

Reply:
- Changing this to a bare struct, I don't see the value in a unit-tuple here.
- Appending the `Error` suffix to the name as you suggest. 
- Changing `NotA` prefix to `Invalid` as I find single letters break readability in camel case. 

Actually, as a general comment, I believe camel case should also apply to acronyms (such as `DNS` -> `Dns`). I know it is tempting to keep full capitalization but years of Java have thought me that since capital letters are the only thing separating words in camel case, it's overall easier to read and more consistent to do so, however shocking at first. It's of course too late to change this in this crate, compatibility and convention are more important than naming nitpicks :)